client/components/ReportsPage/index.jsx

```diff
@@ -0,0 +1,297 @@
+                                tableRows.push(
+                                    <td
+                                        key={`data-${run.id}-${run.apg_example_name}-${techPairHeaders[i].key}`}
+                                    >
+                                        -
+                                    </td>
+                                );
+                            }
+                        }
+                        apgExampleRows.push(
+                            <tr key={`${run.id}-${run.apg_example_name}`}>
+                                {tableRows}
+                            </tr>
+                        );
+                    } else {
+                        // If the row already exists, update the column with percent
+                        let apgExampleRow = apgExampleRows[apgRowIndex];
+                        let dataEntries = [];
+                        for (let dataEntry of apgExampleRow.props.children) {
+                            if (
+                                dataEntry.key.includes(run.at_name) &&
+                                dataEntry.key.includes(run.browser_name)
+                            ) {
+                                dataEntries.push(
+                                    <td key={dataEntry.key}>
+                                        <ProgressBar
+                                            now={percentage}
+                                            variant="info"
+                                            label={`${percentage}%`}
+                                        />
+                                    </td>
+                                );
+                            } else {
+                                dataEntries.push(dataEntry);
+                            }
+                        }
+                        apgExampleRows[apgRowIndex] = (
+                            <tr key={`${run.id}-${run.apg_example_name}`}>
+                                {dataEntries}
+                            </tr>
+                        );
+                    }
+                });
+        }
+
+        return (
+            <Fragment>
+                <Helmet>
+                    <title>{`ARIA-AT Reports`}</title>
+                </Helmet>
+                <h1>Reports Page</h1>
+                <Table bordered hover>
+                    <thead>
+                        <tr>
+                            <th>Design Pattern Examples</th>
+                            {techPairHeaders}
+                        </tr>
+                    </thead>
+                    <tbody>{apgExampleRows}</tbody>
+                </Table>
+            </Fragment>
+        );
+    }
+}
+
+ReportsPage.propTypes = {
+    dispatch: PropTypes.func,
+    publishedRunsById: PropTypes.object
+};
+
+const mapStateToProps = state => {
+    const { publishedRunsById } = state.runs;
+    return { publishedRunsById };
+};
+
+export default connect(mapStateToProps)(ReportsPage);
```

Comment:
I feel really overwhelemed by the amount of (un-tested) logic in this component. And it is only slated to grow as we add more to this page. 

You are doing a lot of munging here that I think really belongs in a utitly and ought to be unit tested. 

I think we have the time aloted to get the reports page correct. And we are plan for this page to go through many iternations so I think its worth taking the time to do this now.

Reply:
+1 to this! I'll add tests in a follow up!